### PR TITLE
Clmm prices

### PR DESCRIPTION
**Pull request by oeble**

### CODE REVIEW

File: programs/scope/Cargo.toml [hunk 1]
```diff
@@ -27,32 +27,35 @@ mainnet = []
...
     "mainnet",
 ], optional = true }
-mpl-token-metadata = { version = "1.13" }
 # Uncomment the line below if you do not have access to the yvaults repo
 #yvaults = { path = "../yvaults_stub", package = "yvaults_stub", optional = true }
-bytemuck = { version = "1.4.0", features = ["min_const_generics", "derive"] }
 sha2 = "0.10.0"
-num-traits = "0.2.15"
-num-derive = "0.4.1"
-spl-associated-token-account = "1.0.3"
```
[the remark this answers is not in the document]
Comment (oeble) [reply programs/scope/Cargo.toml:55]: I made some cleanups as I wanted to move to workspace dependencies but anchor does not support it in 0.28

File: programs/scope/Cargo.toml [hunk 2]
```diff
@@ -62,3 +65,4 @@ solana-program-test = "~1.16.18"
 solana-sdk = "~1.16.18"
 thiserror = "1.0.38"
 test_cpi_caller = { path = "../test_cpi_caller" }
+test-case = "3.2.1"
```
Comment (silviutroscot) [inline programs/scope/Cargo.toml:68]: nit: why not use the latest version, 3.3.0?

File: programs/scope/tests/tests_security_refresh_list.rs [hunk 3]
```diff
@@ -1253,7 +1297,109 @@ async fn test_refresh_list_jlp_wrong_mint() {
...
+    };
+
+    ctx.send_transaction(&[ix]).await.unwrap();
+    let prices: OraclePrices = ctx.get_zero_copy_account(&feed.prices).await.unwrap();
+    // price not updated
+    assert_eq!(prices.prices[TEST_ORCA_ATOB.token].last_updated_slot, 0);
+    assert_eq!(prices.prices[TEST_ORCA_ATOB.token].price.value, 0);
+}
+
+// - [x] Wrong Orca Whirlpool additional token mint B
```
Comment (silviutroscot) [inline programs/scope/tests/tests_security_refresh_list.rs:1357]: shouldn't we have the same tests for Raydium as well?

File: programs/scope/tests/tests_security_refresh_list.rs [hunk 4]
```diff
@@ -42,16 +42,50 @@ const TEST_JLP_ORACLE: OracleConf = OracleConf {
...
+
+const TEST_ORCA_BTOA: OracleConf = OracleConf {
+    pubkey: pubkey!("SomeorcaPriceAccount21111111111111111111111"),
+    token: 7,
+    price_type: TestOracleType::OrcaWhirlpool(false),
+    twap_enabled: false,
+    twap_source: None,
+};
+
+const TEST_RAYDIUM_ATOB: OracleConf = OracleConf {
```
Comment (silviutroscot) [inline programs/scope/tests/tests_security_refresh_list.rs:61]: I think we should also have TEST_RAYDIUM_BTOA

File: programs/scope/Cargo.toml (hunk 2 above)
Comment (oeble) [reply programs/scope/Cargo.toml:68]: it was released 2hours ago 🤷🏼

File: programs/scope/tests/tests_security_refresh_list.rs (hunk 4 above)
Comment (oeble) [reply programs/scope/tests/tests_security_refresh_list.rs:61]: The only difference is the call to sqrt price passing the boolean to false, and this is tested elsewhere. Orca is different because you need to pass the mints and I wanted to ensure no mixup.

File: programs/scope/tests/tests_security_refresh_list.rs (hunk 3 above)
Comment (oeble) [reply programs/scope/tests/tests_security_refresh_list.rs:1357]: No raydium don't need the mints, the mints decimal are stored in the pool account for raydium.

**Review by silviutroscot — APPROVED** [review]

File: off_chain/scope-cli/src/oracle_helpers/orca_whirlpool.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    fn get_number_of_extra_accounts(&self) -> usize {
+        2
+    }
+
+    fn get_mapping_account(&self) -> Option<Pubkey> {
+        Some(self.mapping)
+    }
+
+    async fn get_extra_accounts(&self, _rpc: Option<&dyn AsyncClient>) -> Result<Vec<Pubkey>> {
```
Comment (y2kappa) [inline off_chain/scope-cli/src/oracle_helpers/orca_whirlpool.rs:83]: why is this necessary?

File: programs/scope/src/oracles/mod.rs [hunk 6]
```diff
@@ -146,6 +160,14 @@ where
...
             panic!("DeprecatedPlaceholder is not a valid oracle type")
         }
+        OracleType::OrcaWhirlpoolAtoB => {
+            orca_whirlpool::get_price(true, base_account, clock, extra_accounts)
+        }
+        OracleType::OrcaWhirlpoolBtoA => {
+            orca_whirlpool::get_price(false, base_account, clock, extra_accounts)
+        }
+        OracleType::RaydiumAmmV3AtoB => raydium_ammv3::get_price(true, base_account, clock),
+        OracleType::RaydiumAmmV3BtoA => todo!(),
```
Comment (y2kappa) [inline programs/scope/src/oracles/mod.rs:170]: why?

File: programs/scope/src/oracles/orca_whirlpool.rs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        .ok_or(ScopeError::AccountsAndTokenMismatch)?;
+    let mint_token_b_account_info = extra_accounts
+        .next()
+        .ok_or(ScopeError::AccountsAndTokenMismatch)?;
+
+    // Load main account
+    let pool_data: Whirlpool = account_deserialize(pool)?;
+
+    // Check extra accounts pubkeys
+    require_keys_eq!(
```
Comment (y2kappa) [inline programs/scope/src/oracles/orca_whirlpool.rs:32]: I don't think we need this extra check? the orca pool is definitely a source of truth for the mints

Comment (y2kappa) [reply programs/scope/src/oracles/orca_whirlpool.rs:32]: ah ok you need these two accounts anyway for the decimals
